### PR TITLE
Tmedia 671 caption component

### PR DESCRIPTION
**Pull request by JackHowa**
Ticket: https://arcpublishing.atlassian.net/jira/software/c/projects/TMEDIA/boards/875?modal=detail&selectedIssue=TMEDIA-671&assignee=5e387d6a1b1d910e5dfd7a9b

# Ac: 

Create a <Caption> component

Uses props of Title, Caption and Credit that can be hidden at the block level

May have HTML in it which should render (i.e. `<b>, <i>`)

Have Scss file that adds the hooks to the Sass library

Includes three component properties: Title, Caption, and Credit

Caption - Title

Caption - Caption

Caption - Credit

Create a utility function to handle formatting credits that can be used by consumers - but not built into this component - [engine-theme-sdk/index.tsx at arc-themes-release-version-1.20 · WPMedia/engine-theme-sdk](https://github.com/WPMedia/engine-theme-sdk/blob/arc-themes-release-version-1.20/src/components/ImageMetadata/index.tsx#L47) 

Have test coverage

Have storybook documentation

# Test: 

1. Run `npm run storybook`
2. Go to Caption component 
3. See effect of html passed in via story `Renders html passed into title or caption` 
4. Look at how credit can take in multiple affiliations and/or bylines with stories

## Feedback 

- @nschubach's suggestion of using `.toHaveClass` was not doable as there was no way for me to select figcaption without role or without any text https://github.com/WPMedia/arc-themes-components/pull/29#pullrequestreview-896923800
- Removed out of scope sanitize if security is not a concern https://github.com/WPMedia/arc-themes-components/pull/29#discussion_r819742123
- Used dangerously set html based on feedback and requirements https://github.com/WPMedia/arc-themes-components/pull/29#discussion_r817198643
- Naming standards discussion from before

### CODE REVIEW

File: src/components/caption/index.jsx [hunk 1]
```diff
@@ -0,0 +1,35 @@
+import PropTypes from "prop-types";
+import formatCredits from "../../utils/format-credits";
+
+const COMPONENT_CLASS_NAME = "c-caption";
+
+const Caption = ({ caption, className, credit, title }) => {
+	const formattedCredit = formatCredits(credit);
```
Comment (matthewroach) [inline src/components/caption/index.jsx]: It was discussed that this utility would be for consumers to use and not directly built into the component

**Review by matthewroach — DISMISSED** [review]

File: src/components/caption/index.test.jsx [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+			<Caption
+				title="title"
+				credit={{ by: [{ name: "Joe Doe" }], affiliation: [{ name: "Stock Photo" }] }}
+				caption="caption"
+			/>
+		);
+
+		expect(screen.queryByText("title")).not.toBeNull();
+		// expect(screen.queryByText({{ by: [{ name: "bbbyyy" }], affiliation: [{ name: "affff" }] }})).not.toBeNull();
+		// expect(screen.queryByText("caption")).not.toBeNull();
```
Comment (nschubach) [inline src/components/caption/index.test.jsx]: Remove?

File: src/components/caption/index.jsx [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+
+	return title || caption || formattedCredit ? (
+		<figcaption
+			className={className ? `${COMPONENT_CLASS_NAME} ${className}` : `${COMPONENT_CLASS_NAME}`}
+		>
+			{title ? (
+				<span dangerouslySetInnerHTML={{ __html: `${title} ` }} className="c-caption__title" />
```
Comment (nschubach) [inline src/components/caption/index.jsx]: Should be perhaps use the const for this?
```suggestion
				<span dangerouslySetInnerHTML={{ __html: `${title} ` }} className=`${COMPONENT_CLASS_NAME}__title` />
```

File: src/components/caption/index.jsx [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+			) : null}
+			{caption ? (
+				<span dangerouslySetInnerHTML={{ __html: `${caption} ` }} className="c-caption__caption" />
```
Comment (nschubach) [inline src/components/caption/index.jsx]:
```suggestion
				<span dangerouslySetInnerHTML={{ __html: `${caption} ` }} className=`${COMPONENT_CLASS_NAME}__caption` />
```

File: src/components/caption/index.jsx [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+			) : null}
+			{formattedCredit ? <span className="c-caption__credit">{formattedCredit}</span> : null}
```
Comment (nschubach) [inline src/components/caption/index.jsx]:
```suggestion
			{formattedCredit ? <span className=`${COMPONENT_CLASS_NAME}__credit`>{formattedCredit}</span> : null}
```

File: src/components/caption/index.stories.mdx [hunk 6]
```diff
@@ -0,0 +1,168 @@
+import { Meta, Story, Preview, Props } from "@storybook/addon-docs";
```
Comment (nschubach) [inline src/components/caption/index.stories.mdx]: `Preview` and `Props` were renamed to `Canvas` and `ArgsTable` respectively.  Can you update this?

**Comment by brentmiller1973** [comment]
@JackHowa This is very nitpicky, but in story examples where we demonstrate that HTML can be embedded, using divs or breaks are not immediately apparent.  `<b>` or `<strong>` tags give that feedback better to the user.  What do you think to modifying .storybook/preview-head.html to inject some styles?  So for `<b` just update the file to:

```
<!-- .storybook/preview-head.html -->
<style>
    b {
        font-weight: bold;
    }
</style>
```

Then you will get a better demonstration in the "Renders html passed into title or caption" story:

image

Or as an alternative, just inline the style in the story:

```
<Preview>
	<Story name="Renders html passed into title or caption">
		<Caption
			title='<p><b style="font-weight:bold">Important Alert</b></p>'
			caption="There is man on the loose biting dogs</div><div>All dogs should evacuate the area"
			credit={{ by: [{ name: "John Doe" }], affiliation: [{ name: "Stock Photo" }] }}
		/>
	</Story>
</Preview>
```

**Review by nschubach — DISMISSED** [review]
Class requests, Storybook Doctag name updates, and some commented code to remove.

**Comment by matthewroach** [comment]
@visionbegin - To your comment about setting default values for bold. We should do this inside of the Sass library, maybe we are resetting too much. Think like `<b>` and `<em>` should have the bold and italic CSS set out of the box - why should we remove it? or even make additional work for ourselves to have to redefine it?

File: src/components/caption/index.jsx (hunk 1 above)
Comment (JackHowa) [reply src/components/caption/index.jsx]: updated!

File: src/components/caption/index.test.jsx (hunk 2 above)
Comment (JackHowa) [reply src/components/caption/index.test.jsx]: updated

**Comment by brentmiller1973** [comment]
> @visionbegin - To your comment about setting default values for bold. We should do this inside of the Sass library, maybe we are resetting too much. Think like `<b>` and `\<em\>` should have the bold and italic CSS set out of the box - why should we remove it? or even make additional work for ourselves to have to redefine it?

@matthewroach I'm not thinking so much about our Sass settings as I am thinking about providing immediate feedback to the client visually in the story.  While my initial suggestion was to modify the storybook style header, I still think there is value to inline the styles on some stories (especially html injection) to demonstrate this.

**Comment by JackHowa** [comment]
is this new? `npm run storybook`

```
37% building 230/243 modules 13 active ...s/node_modules/core-js/modules/esnext.symbol.metadata.jsWarning: components "paragraph" is configured with no options
    src/scss/root/_index.scss 68:4   token-map-to-css-variable()
    src/scss/root/_index.scss 98:4   root-tokens()
    src/scss/root/_index.scss 132:4  @forward
    src/scss/index.scss 1:1          @forward
    scss.scss 1:1                    @use
    .storybook/theme.scss 1:1        root stylesheet

```

File: src/components/caption/index.stories.mdx (hunk 6 above)
Comment (JackHowa) [reply src/components/caption/index.stories.mdx]: updated! thanks @nschubach

**Comment by JackHowa** [comment]
updated based on feedback!

**Comment by JackHowa** [comment]
Just for reference, the css reset reset the font-style. That's why it doesn't look as expected: 

```
	&,
	&::before,
	&::after {
		font: inherit;
```

cc @visionbegin 
Screen Shot 2022-03-04 at 15 46 29

**Comment by JackHowa** [comment]
I guess this gets back to the discussion of adding "example" or something themes to the storybook picker. I'm not sure if this conversation is relevant to this pr

**Review by matthewroach — DISMISSED** [review]
The separate util function looks good, but there's no documentation for it that's accessible to developers. We can add stories for these utils too, and use the grouping feature in storybook to put them under a Utilities grouping

**Comment by matthewroach** [comment]
> is this new? `npm run storybook`
> 
> ```
> 37% building 230/243 modules 13 active ...s/node_modules/core-js/modules/esnext.symbol.metadata.jsWarning: components "paragraph" is configured with no options
>     src/scss/root/_index.scss 68:4   token-map-to-css-variable()
>     src/scss/root/_index.scss 98:4   root-tokens()
>     src/scss/root/_index.scss 132:4  @forward
>     src/scss/index.scss 1:1          @forward
>     scss.scss 1:1                    @use
>     .storybook/theme.scss 1:1        root stylesheet
> ```

This is due to the `.storybook/theme.scss` file having a reference for `paragraph: ()` and nothing set - the warnings is built into our Sass library to provide developers additional feedback

**Comment by JackHowa** [comment]
> The separate util function looks good, but there's no documentation for it that's accessible to developers. We can add stories for these utils too, and use the grouping feature in storybook to put them under a Utilities grouping

updated @matthewroach

**Comment by JackHowa** [comment]
> > is this new? `npm run storybook`
> > ```
> > 37% building 230/243 modules 13 active ...s/node_modules/core-js/modules/esnext.symbol.metadata.jsWarning: components "paragraph" is configured with no options
> >     src/scss/root/_index.scss 68:4   token-map-to-css-variable()
> >     src/scss/root/_index.scss 98:4   root-tokens()
> >     src/scss/root/_index.scss 132:4  @forward
> >     src/scss/index.scss 1:1          @forward
> >     scss.scss 1:1                    @use
> >     .storybook/theme.scss 1:1        root stylesheet
> > ```
> 
> This is due to the `.storybook/theme.scss` file having a reference for `paragraph: ()` and nothing set - the warnings is built into our Sass library to provide developers additional feedback

updated this as well to remove the warning

**Review by nschubach — APPROVED** [review]
Looks like all the concerns were addressed and I do not see any issue with the storybook or test coverage.

**Pull request merged**